Dashboard: Variable Referencing in Lattice Configuration

proy30 opened the pull request:
Supersedes #829 due to changes in a different branch.

--------------
This PR allows dashboard users to define variable names, assign numeric values, and reference these variables in lattice configurations.

- This PR validates a variable name by using Python’s built-in isidentifier() method and the keyword library to ensure it starts correctly, contains only allowed characters, isn’t a reserved word, and is unique among all variables.


### Clarification on `ui_input` vs `sim_input` behavior introduced in this PR

- If the user inputs `x` in `parameter1`, and `x` **is defined** as a variable with value `50`:
  - `ui_input = 'x'`
  - `sim_input = 50`

- If the user inputs `x` in `parameter1`, and `x` **is not defined**:
  - `ui_input = 'x'`
  - `sim_input = 'x'` *(not `''`)*  
    The simulation is disabled in this case if the `param_type` is not `str`.

- If the user inputs `7` directly in `parameter1`:
  - `ui_input = 7`
  - `sim_input = 7`

proy30 committed:
add variables tab, some ui components, some funcitonality, basic validation

proy30 committed:
add validation to require first char of var name as alphabetical letter

proy30 committed:
do not save var_name if validation fails

proy30 committed:
show 'add_variable' button only on last index

doesnt make sense to display the button on all rows if it adds the button in the same location

proy30 committed:
restructure vlaidation

proy30 committed:
add variable reset button

proy30 committed:
Restructure the code

proy30 committed:
Update docstring and type hints

proy30 committed:
add v_model for correct flushing

proy30 committed:
dont run sim if error in variables

proy30 committed:
cleaup

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

proy30 committed:
Merge branch 'development' into dashboard/duplicate_var

proy30 committed:
update: defaults on card button component

proy30 committed:
convert vue2 to vue3 logic and switch to using existing components

proy30 committed:
separate lattice variable config logib to new file

proy30 committed:
Make text fields work in vue3 and add clearable property

proy30 committed:
fix: v_model

proy30 committed:
remove: unused code

proy30 committed:
solve reactivity issues

The UI will now properly update to user expectations. For example if a user enters a non-existing variable name first, then enters the variable name in the config and a valid value, the lattice list config will automatically interpret it. Previously, this same case would have required the user to reenter the var name (ie. AFTER the variable was made in the config)

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

proy30 committed:
fix: CodeQL and some small cleanup changes

proy30 committed:
separate lattice config setting tabs logic

proy30 committed:
simplify

proy30 committed:
Update conditions to ensure a valid python variable name is given

using keyword library

proy30 committed:
update: variable name validation

proy30 committed:
remove: print statements

proy30 committed:
Merge branch 'dashboard/variable_referencing_in_lattice_config' of github.com:proy30/impactx into dashboard/variable_referencing_in_lattice_config

proy30 commented:
https://github.com/user-attachments/assets/c48bb282-2c6e-4ef9-b983-cf13aa61d045

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

proy30 committed:
add staticmethod tag

proy30 committed:
Merge branch 'dashboard/variable_referencing_in_lattice_config' of github.com:proy30/impactx into dashboard/variable_referencing_in_lattice_config

cemitch99 reviewed (approved):
It looks good to me.  Thanks a lot.

EZoni commented on src/python/impactx/dashboard/Input/latticeConfiguration/latticeMain.py:
```diff
@@ -170,17 +154,74 @@ def on_add_lattice_element_click():
+    lattice_variable, variable_index = (
+        LatticeVariableHandler.determine_if_existing_variable(ui_value)
+    )
+    potentially_lattice_variable = LatticeVariableHandler.is_valid_variable_name(
+        ui_value
+    )
```
@proy30 

I'm starting to review this PR, I will add single comments here and there as I go through it.

My understanding, based on the implementation of `is_valid_variable_name`, is that `ui_value` here is a `str` that represents the name of the "helper" variable used to encode the value of a parameter that ImpactX and/or the dashboard know about. Is that right? If this is the case, I wonder if `value` is the right name for this variable. If the "helper" variable is called `n` and its value is set to `10` (and, say, used to set the value of several parameters that ImpactX and/or the dashboard know about), then what does `ui_value` represent in this case, the name `n` or the value `10`?

proy30 replied:
In your example, `ui_value` would be `n` and `sim_value` would be `10`.

That said, `ui_value` isn’t always a `str` (variable name). If a user directly enters `7` into a text field, then `ui_value` would be `7`, since that’s what’s shown in the UI.

So, whatever is visible to or entered by the user — whether it’s a variable name or a number — ends up as the `ui_value`. The `sim_value` is the actual value used in the simulation (either the variable’s value or the raw input).

As for naming, I’m open to suggestions. I went with `ui_value` since it can hold both a number or a string, but maybe something like `ui_input` could be clearer?

EZoni replied:
Understood, thanks. Just to clarify, before I possibly make naming suggestions, `ui_value` can be either a `str` or the numerical value of the quantity represented by the `str`, while `sim_value` is always a numerical value? Or can you have a situation where both are `str` (maybe what you referred to as "raw input" in your comment above)?

proy30 replied:
I was about to say that `sim_value` is always a numerical value — but after checking that’s not true. A user could type something like `x` into a parameter, even though `x` hasn’t been defined as a variable yet. In that case, `sim_value` would be `x`, a `str`.

Right now, the dashboard handles this by throwing an error message and disabling the simulation. We could also make it so that if `sim_value` is a `str`, we just set it to an empty string `""`.

**I believe I actually tried this approach in the beginning of this PR,** but something didn’t work as expected — I think some reactivity issue --  so I left it as-is, especially since we already block the sim when a non-numerical input is given where a number is expected.

With that said, `sim_value` might need a variable rename then.

EZoni replied:
Okay, I think throwing an error and disabling the simulation is fine, especially if this is already the general behavior anyways. I'm thinking that maybe your suggestion of using `input` instead of `value` may be a good one, like `ui_input` and `sim_input`. What do you think?

proy30 replied:
Yes, I agree that is a better naming.

Also for clarification:

If user inputs `x` in `parameter1` and `x` is defined as a variable with value `50`:
`ui_value` = `'x'`
`sim_value` = `50`

If user inputs `x` in `parameter1` and `x` is not defined as a variable:
`ui_value` = `'x'`
`sim_value` = `'x'` instead of `sim_value` = `''`

If user directly inputs `7`  in `parameter1`:
`ui_value` = `7`
`sim_value` = `7`

EZoni replied:
Perfect, would you maybe add this example that you just made to the PR description as well, so it's a little more visible if one looks back at the PR in the future? Thanks!

And/or add something similar as inline comment in the code as well.

EZoni commented on src/python/impactx/dashboard/Input/generalFunctions.py:
Looking generally at the function `update_simulation_validation_status`, is `error_details`, which in this PR you append with what is returned by `has_error_in_variables`, used elsewhere other than for disabling the simulation button? If not, would it make sense to maybe print `error_details` to standard output or something similar? Or did you see this more like an internal debugging tool for developers only?

proy30 replied:
The error details aren’t shown directly to the user on the dashboard since users already see error messages on the specific input fields.

We could refactor this to use a simple true/false flag instead, but I originally kept the detailed messages for debugging during the early stages of the dashboard and just never updated it after that.

I do know that `update_simulation_validation_status` is messy and have plans to refactor it to reduce the amount of manual print statements to `error_details`.

EZoni replied:
Understood, thanks. Sounds like work that could be done in future PRs, without urgency.

EZoni commented on src/python/impactx/dashboard/Input/latticeConfiguration/variable_handler.py:
```diff
@@ -0,0 +1,270 @@
+        new_variable = {key: "" for key in state.variables[0]}
+        state.variables.append(new_variable)
+        state.dirty("variables")
+        LatticeVariableHandler.update_delete_availability()
```
Could this manual call be removed if `update_delete_availability` is decorated with `@state.change("variables")`?
```suggestion
```

proy30 replied:
I originally had this handled manually instead of using `@state.change("variables")` because that decorator triggers on **any** change to the `variables` list — including edits to names or values — not just when the number of variables changes.

I think the performance impact is minimal, so I’m not opposed to switching to the reactive approach for simplicity.

EZoni replied:
Understood, makes sense. Yes, I was thinking that moving the availability control to a function that listens to changes in `variables` (i.e., all changes), as you did in your last commit, may make things more robust (thinking of someone other than you who may make changes to the code in the future, without necessarily mastering the whole infrastructure in detail), as it reduces the need for ad-hoc manual calls in exchange for the overhead that you mentioned.

EZoni commented on src/python/impactx/dashboard/Input/latticeConfiguration/variable_handler.py:
```diff
@@ -0,0 +1,270 @@
+
+        state.variables.pop(index)
+        state.dirty("variables")
+        LatticeVariableHandler.update_delete_availability()
```
Could this manual call be removed if `update_delete_availability` is decorated with `@state.change("variables")`?
```suggestion
```

EZoni commented on src/python/impactx/dashboard/Input/latticeConfiguration/variable_handler.py:
```diff
@@ -0,0 +1,270 @@
+            {"name": init_value, "value": init_value, "error_message": init_value}
+        ]
+        state.dirty("variables")
+        LatticeVariableHandler.update_delete_availability()
```
Could this manual call be removed if `update_delete_availability` is decorated with `@state.change("variables")`?
```suggestion
```

proy30 committed:
refactor: move 'update_delete_availability' to @state.change(variables)

simplifies the manual call, however now it will be called any time a change to variable name/value happens

proy30 committed:
update: variable name for 'sim_value' and 'ui_value'

to 'sim_input' and 'ui_input' respectively
